src/Prefetch.cpp

```diff
@@ -0,0 +1,348 @@
+
+        string pstr = std::to_string(ptmp++);
+        string varname_prefetch_buf = varname + "_prefetch_" + pstr + "_buf";
+        Expr var_prefetch_buf = Variable::make(Int(32), varname_prefetch_buf);
+
+        // Establish the variables for buffer strides, box min & max
+        vector<Expr> stride_var(dims);
+        vector<Expr> extent_var(dims);
+        vector<Expr> min_var(dims);
+        vector<Expr> max_var(dims);
+        for (int i = 0; i < dims; i++) {
+            string istr = std::to_string(i);
+            // string extent_name = varname + ".extent." + istr;
+            string stride_name = varname + ".stride." + istr;
+            string extent_name = varname + ".extent." + istr;
+            string min_name = varname + "_prefetch_" + pstr + "_min_" + istr;
+            string max_name = varname + "_prefetch_" + pstr + "_max_" + istr;
+
+            stride_var[i] = Variable::make(Int(32), stride_name);
+            extent_var[i] = Variable::make(Int(32), extent_name);
+            min_var[i] = Variable::make(Int(32), min_name);
+            max_var[i] = Variable::make(Int(32), max_name);
+
+            // Record let assignments
+            // except for stride, already defined elsewhere
+            plets.push_back(make_pair(min_name, box[i].min));
+            plets.push_back(make_pair(max_name, box[i].max));
+        }
+
+        // This box should not be prefetched
+        if (!do_prefetch) {
+            debug(dbg_prefetch1) << "  Info: not prefetching " << varname << "\n";
+            return;
+        }
+
+        // Create a buffer_t object for this prefetch.
+        vector<Expr> args(dims*3 + 2);
+
+        Expr first_elem = Load::make(t, varname, 0, BufferPtr(), Parameter());
+        args[0] = Call::make(Handle(), Call::address_of, {first_elem}, Call::PureIntrinsic);
+        args[1] = make_zero(t);
+        for (int i = 0; i < dims; i++) {
+            args[3*i+2] = min_var[i];
+            args[3*i+3] = max_var[i] - min_var[i] + 1;
+            args[3*i+4] = stride_var[i];
+        }
+
+        // Create the create_buffer_t call
+        Expr prefetch_buf = Call::make(type_of<struct buffer_t *>(), Call::create_buffer_t,
+                              args, Call::Intrinsic);
+        plets.push_back(make_pair(varname_prefetch_buf, prefetch_buf));
+
+        // Create the prefetch call
+        Expr num_elem = stride_var[dims-1] * extent_var[dims-1];
+        vector<Expr> args_prefetch(4);
+        args_prefetch[0] = dims;
+        args_prefetch[1] = elem_size_bytes;
+        args_prefetch[2] = num_elem;
+        args_prefetch[3] = var_prefetch_buf;
+        Stmt stmt_prefetch = Evaluate::make(Call::make(Int(32), Call::prefetch_buffer_t,
+                              args_prefetch, Call::Intrinsic));
+        // TODO: Opt: Keep running sum of bytes prefetched on this sequence?
+        // TODO: Opt: Keep running sum of number of prefetch instructions issued
```

Comment:
I think the only reasonable thing we can do with this information is emit debug warnings that the amount being prefetched, or the number of prefetches is not useful (thrashing the cache). I think this code belongs in the runtime, and there is already some of it there, so I think these TODOs should be removed (or moved to the runtime).